File: session-01/task-04/word-index.js
const fs = require('fs');

const LINES_PER_PAGE = 45;
const MAX_SIZE_LINE = 80;
const STOP_FREQUENCY_LIMIT = 100;

function removePunctuation(word) {
    word = word.replace(",", "");
    word = word.replace("\r", "");
    if(word != "e.g." && word != "i.e.")
        word = word.replace(".", "");

    return word;
}

module.exports = function(file_path) {
    var word_count = new Map();
    var word_index = new Map();
    var line_count = 0;
    var page_number = 1;

    var data = fs.readFileSync(file_path, "utf8");
    
    var line = "";
    var lines = [];
    var line_size_count = 0;
    for (var i = 0; i < data.length; i++) {
        if (line_size_count == MAX_SIZE_LINE || data.charAt(i)=="\n") {
            lines.push(line);
            line = "";
            line_size_count = 0;

            if(data.charAt(i)=="\n")
                continue;
        }
        
        line_size_count++;
        line = line + data.charAt(i);
    }
    lines.push(line);

    lines.forEach(line => {
        line_count++;

        var words = line.split(" ");
        words.forEach(word => {
            word = removePunctuation(word);
            if(word.length==0) 
                return;
            
            if(!word_count.has(word)) {
                word_count.set(word, 0);
                word_index.set(word, new Set());
            }
            word_count.set(word, word_count.get(word) + 1);
            word_index.get(word).add(page_number);
        });

        if(line_count==LINES_PER_PAGE) {
            line_count = 0;
            page_number++;
        }
    });

    word_count.forEach((count, word) => {
        if (count > STOP_FREQUENCY_LIMIT)
            word_index.delete(word);
    });
    
    output = "";
    var words = Array.from(word_index.keys());
    words.sort((a,b) => {
        if(a.toLowerCase() < b.toLowerCase()) return -1;
        if(b.toLowerCase() < a.toLowerCase()) return 1;
        return 0;
    })
    words.forEach(word => {
        output += word + ": ";
        word_index.get(word).forEach(i => {
            output += i + ", ";
        });
        output += "\n";
    });
    console.log(output);

    return word_index;
}

File: session-01/task-02/word_index.py
#!/usr/bin/env python

# Program Description:
#
# Word Index is a program that takes a plain text file as input and
# outputs all the words contained in it
# sorted alphabetically along with the page numbers on which they occur.
# The program assumes that a page is a
# sequence of 45 lines, each line has max 80 characters, and there is no
# hyphenation. Additionally, Word Index
# must ignore all words that occur more than 100 times.

####################################
# Free Flow, a.k.a. No Style
####################################

import collections
import string
import sys

# Global "constants"
LINES_PER_PAGE       =  45
MAX_SIZE_LINE        =  80
STOP_FREQUENCY_LIMIT = 100


# Defining a main method makes testing easier
def main(file_path):
    word_count = {}
    word_index = {}
    line_count = 0
    page_number = 1

    file_handle = open(file_path, "r")
    while True:
        line = file_handle.readline(MAX_SIZE_LINE)
        if not line:
            break

        line_count += 1

        words = line.split()
        for word in words:
            word = word.replace(',', '')
            if word != 'e.g.' and word != 'i.e.':
                for punctuation_char in string.punctuation:
                    word = word.replace(punctuation_char, '')

            if word == '':
                continue

            if word not in word_count.keys():
                word_count[word] = 0
                word_index[word] = []
            word_count[word] = word_count[word] + 1

            if page_number not in word_index[word]:
                word_index[word].append(page_number)

        if line_count == LINES_PER_PAGE:
            line_count = 0
            page_number = page_number + 1

    file_handle.close()

    for word, count in word_count.items():
        if count > STOP_FREQUENCY_LIMIT:
            word_index.pop(word)

    for word, index in collections.OrderedDict(sorted(word_index.items(), key=lambda i: i[0].casefold())).items():
        print(word + ": " + ", ".join(map(str, index)))

    return word_index


if __name__ == "__main__":
    main(sys.argv[1])
